### PR TITLE
Fix a bug when finding calls with one or more arguments

### PR DESCRIPTION
**Pull request by tomokinakamaru**
Before this fix: `foo(_, ...)` matches `foo()`, `foo(1)`, and `foo(1, 2)`
After this fix: `foo(_. ...)` matches only `foo(1)` and `foo(1, 2)`

### CODE REVIEW

File: src/pattern/node/call.ts [hunk 1]
```diff
@@ -20,41 +20,58 @@ export class Call extends Node {
...
+    const pattern1 = getPatternNodeAt(patternNodes, 1) as Node;
+    if (pattern1 === undefined) {
+      return true;
     }
-    return argIdx === this.args.length && exprIdx === exprs.length;
+    if (pattern1.match(typescript0, typeChecker)) {
+      return matchArgs(patternNodes.slice(2), typescriptNodes.slice(1), typeChecker);
+    }
+    return matchArgs(patternNodes, typescriptNodes.slice(1), typeChecker);
+  }
```
Comment (wata727) [inline src/pattern/node/call.ts]: This is very complicated... 

A naive idea would be to implement a function that takes a pair of nodes and checks whether each node is matched, but we know that in patterns where there is a node after ellipsis it does not work as expected. I have no ideas :man_shrugging:

**Comment by wata727** [comment]
BTW, I noticed the following somewhat strange behavior:

```
$ bin/tyscan.js console ~/ghq/github.com/wata727/sideci/frontend                                                              [11:33:58]
TyScan console

Available commands:
  - find <pattern>  Find <pattern>
  - reload          Reload TypeScript files
  - exit            Exit

> find middleware
/home/wata727/ghq/github.com/wata727/sideci/frontend/src/cheshire/store.ts#L25C53       middleware
/home/wata727/ghq/github.com/wata727/sideci/frontend/src/cheshire/routing/__tests__/Link.test.tsx#L69C94        middleware
/home/wata727/ghq/github.com/wata727/sideci/frontend/src/cheshire/routing/__tests__/React.test.tsx#L92C94       middleware
/home/wata727/ghq/github.com/wata727/sideci/frontend/src/cheshire/routing/__tests__/React.test.tsx#L114C94      middleware
/home/wata727/ghq/github.com/wata727/sideci/frontend/src/cheshire/routing/__tests__/Router.test.ts#L110C31      middleware
/home/wata727/ghq/github.com/wata727/sideci/frontend/src/cheshire/routing/__tests__/Router.test.ts#L112C21      middleware
/home/wata727/ghq/github.com/wata727/sideci/frontend/src/cheshire/routing/__tests__/Router.test.ts#L134C32      middleware
/home/wata727/ghq/github.com/wata727/sideci/frontend/src/cheshire/routing/__tests__/Router.test.ts#L170C32      middleware
/home/wata727/ghq/github.com/wata727/sideci/frontend/src/cheshire/routing/__tests__/Router.test.ts#L204C32      middleware
/home/wata727/ghq/github.com/wata727/sideci/frontend/src/cheshire/routing/__tests__/Router.test.ts#L241C32      middleware
/home/wata727/ghq/github.com/wata727/sideci/frontend/src/cheshire/routing/__tests__/Router.test.ts#L270C32      middleware
/home/wata727/ghq/github.com/wata727/sideci/frontend/src/cheshire/routing/__tests__/Router.test.ts#L298C32      middleware
/home/wata727/ghq/github.com/wata727/sideci/frontend/src/cheshire/routing/__tests__/Router.test.ts#L327C32      middleware
> find middleware(...)
> 
```

I'm not sure if this behavior is intended, is this not an equivalent pattern?

**Comment by tomokinakamaru** [comment]
> BTW, I noticed the following somewhat strange behavior:
> 
> ```
> $ bin/tyscan.js console ~/ghq/github.com/wata727/sideci/frontend                                                              [11:33:58]
> TyScan console
> 
> Available commands:
>   - find <pattern>  Find <pattern>
>   - reload          Reload TypeScript files
>   - exit            Exit
> 
> > find middleware
> /home/wata727/ghq/github.com/wata727/sideci/frontend/src/cheshire/store.ts#L25C53       middleware
> /home/wata727/ghq/github.com/wata727/sideci/frontend/src/cheshire/routing/__tests__/Link.test.tsx#L69C94        middleware
> /home/wata727/ghq/github.com/wata727/sideci/frontend/src/cheshire/routing/__tests__/React.test.tsx#L92C94       middleware
> /home/wata727/ghq/github.com/wata727/sideci/frontend/src/cheshire/routing/__tests__/React.test.tsx#L114C94      middleware
> /home/wata727/ghq/github.com/wata727/sideci/frontend/src/cheshire/routing/__tests__/Router.test.ts#L110C31      middleware
> /home/wata727/ghq/github.com/wata727/sideci/frontend/src/cheshire/routing/__tests__/Router.test.ts#L112C21      middleware
> /home/wata727/ghq/github.com/wata727/sideci/frontend/src/cheshire/routing/__tests__/Router.test.ts#L134C32      middleware
> /home/wata727/ghq/github.com/wata727/sideci/frontend/src/cheshire/routing/__tests__/Router.test.ts#L170C32      middleware
> /home/wata727/ghq/github.com/wata727/sideci/frontend/src/cheshire/routing/__tests__/Router.test.ts#L204C32      middleware
> /home/wata727/ghq/github.com/wata727/sideci/frontend/src/cheshire/routing/__tests__/Router.test.ts#L241C32      middleware
> /home/wata727/ghq/github.com/wata727/sideci/frontend/src/cheshire/routing/__tests__/Router.test.ts#L270C32      middleware
> /home/wata727/ghq/github.com/wata727/sideci/frontend/src/cheshire/routing/__tests__/Router.test.ts#L298C32      middleware
> /home/wata727/ghq/github.com/wata727/sideci/frontend/src/cheshire/routing/__tests__/Router.test.ts#L327C32      middleware
> > find middleware(...)
> > 
> ```
> 
> I'm not sure if this behavior is intended, is this not an equivalent pattern?

This is intended but I agree that it is confusing. `middleware` is a pattern to find an identifier whose name is `middleware`, and `middleware(...)` is one to find *call to a function* whose name is `middleware`. I think we need to discuss this behavior should be kept or not.

File: src/pattern/node/call.ts (hunk 1 above)
Comment (tomokinakamaru) [reply src/pattern/node/call.ts]: Yes, very complicated so I should revise this part. I will try it now.

Things become much easier if we do not allow patterns like `foo(_: number, ..., _: number)`, a pattern that has `...`  in the middle of the argument sequence.

Comment (tomokinakamaru) [reply src/pattern/node/call.ts]: I revised `matchArgs` and added comments to each step, but unfortunately, it is still complicated... The matching process is kind of the emulation of a state machine so the code would be more clear if I have defined classes for the state machine. However, I think such implementation is too much just for this matching process.

**Pull request merged**